### ISSUE TITLE
No Argument-less Constructor

### ISSUE THREAD

**Issue by gondsm** [JuliaPOMDP/DESPOT.jl#25]
Hello! I'm trying to get to grips with the `DESPOT` solver. I notice that running `solver = DESPOTSolver()` yields a `MethodError: no method matching DESPOT.DESPOTSolver()`.

I've seen some examples where the arguments are defined and the constructor is called, but wouldn't it be nice to have an argument-less default constructor? Would it be unreasonable to expect one, given DESPOT's theoretical framework? Is there no set of "reasonable" arguments that could be used by default?

**Comment by zsunberg** [comment]
Hi @gondsm Yes. This version of DESPOT does take some effort to use. I am currently working on a rewrite. https://github.com/zsunberg/ARDESPOT.jl
The new code is implemented based on the DESPOT journal paper, rather than the C++ code, so I am hoping it will be easier to understand. Currently it works for lambda = 0, but the regularization does not work. It should be fairly easy to make a zero-argument constructor with the new code, but I haven't implemented good default bounds yet.

**Comment by zsunberg** [comment]
Currently, the constructor in ARDESPOT.jl only requires the bounds argument, see, e.g. https://github.com/zsunberg/ARDESPOT.jl/blob/8846c6bb61eb31848c36c6b7a5edb1eb6c908cf3/test/runtests.jl#L35

**Comment by gondsm** [comment]
Hi @zsunberg, thanks for the replies! 

In terms of reliability and accuracy, which of the implementations do you feel is the best at this point?

**Comment by zsunberg** [comment]
I think they are both accurate for lambda=0. Both *definitely* do not work
correctly for lambda>0. ARDESPOT.jl is easier to use, may work for more
problems, and has a few extra debugging tools like pretty printing of
trees. I don't foresee DESPOT.jl being maintained in the future, but I am
actively working on ARDESPOT.jl, so if you are going to contribute, you
should definitely contribute to ARDESPOT.jl. Unfortunately ARDESPOT.jl is
not yet documented.

In short, I'd recommend using ARDESPOT.jl

**Comment by zsunberg** [comment]
ARDESPOT attempts to follow the pseudocode in this paper as closely as
possible http://jair.org/media/5328/live-5328-9824-jair.pdf

**Comment by zsunberg** [comment]
whereas DESPOT.jl was designed to *exactly* emulate this software
https://github.com/AdaCompNUS/despot

**Comment by gondsm** [comment]
Oh, I see. Well, in that case I believe the paper should provide more than enough guidance through the code. I'll be taking a look at ARDESPOT when I get back to the lab tomorrow. Thanks very much! 😁👍

**Comment by gondsm** [comment]
(what I mean is that the lack of docs should be compensated by the paper. Perhaps I'll even contribute some docs!)

**Comment by ebalaban** [comment]
@gondsm, DESPOT constructor does not have any required function arguments, only optional ones (take a look in [solver.jl](https://github.com/JuliaPOMDP/DESPOT.jl/blob/master/src/solver.jl)). You are getting that error for a different reason - it's because what you do need to specify are the **types** of you states, actions, observations, etc. For example, you should be able to say something like this:
```julia
 solver = DESPOTSolver{Int64, Int64, Int64, MyBoundType, RandomStreams}()
```
In the earlier versions of this code `solver = DESPOTSolver()` would have worked fine. The reason I introduced type specification at one point is performance. Fully specifying key types resulted in about a 3x solver speed-up (and key types is the kind of information you should have before instantiating your solver anyway).

This implementation is indeed designed to emulate the C++ code both to ensure correctness and to be able to accurately compare performance to C++. As far as the latter goes, this implementation should be within a factor of 1.5-2.0 of C++. I intend to continue supporting it, but updates may be somewhat delayed, as my workload is rather heavy at the moment.

Hope this helps.

**Comment by gondsm** [comment]
@ebalaban I see, I don't actually need arguments but to specify the types... I can see how that would improve stuff, allowing Julia to compile everything neatly only once. I'm new to Julia, so this stuff still kind of slips past me. Sorry!

In essence, there *is* an argument-less constructor, but I still need to specify the types. That's cleared up! Many thanks, I think I can get stuff to work now!

That answers the original question of the issue, so feel free to close it ;)

**Comment by zsunberg** [comment]
Hey @ebalaban, sorry, I shouldn't have spoken for you about maintaining the package.

In addition to requiring the type arguments, in practice the `DESPOTSolver` constructor requires the `next_state` and `curr_obs` arguments since it is not common for types to have a zero-argument constructor (e.g. `Bool()` and `Int()` throw errors). This is confusing to the user because it is not obvious whether they need to be initialized to some particular value.

Type stability is certainly necessary for good performance, but it is not necessary to know the types at solver construction time. If you push type inference back to solve time instead of requiring it at solver construction time, then the need for all of the arguments mentioned above is removed. This can be accomplished by parameterizing the policy type instead of the solver type. I think making that change, so that the constructor does not require any unnecessary arguments, would be one of the best ways to improve the usability of this package.

**Comment by ebalaban** [comment]
@zsunberg, yes, zero argument constructors can indeed be a problem for some types. Pushing off type inference until policy is constructed is not a bad idea, since we'd have the POMDP to extract the types from at that point. I'll implement that.

@gondsm, no worries at all. I'll close this issue then, but if you have any other questions, please let me know.

**Comment by gondsm** [comment]
@ebalaban Thank you very much!
